src/hotspot/share/opto/gcm.cpp

```diff
@@ -769,6 +773,39 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
     }
   }
   worklist_def_use_mem_states.push(nullptr, initial_mem);
+  Block* initial_mem_block = get_block_for_node(initial_mem);
+  assert(initial_mem_block != nullptr, "sanity");
+  if (load->in(0) != nullptr) {
+    assert(initial_mem_block->dominates(early), "invariant");
+    // If the load has an explicit control input, walk up the dominator tree
+    // from the early block (inclusive) to the initial memory block
```

Comment:
Would be nice to give a motivation for this: why is that all of the relevant blocks?

Reply:
I have updated the comment now, please check if it is clearer now. `initial_mem` and relevant Phi memory aliases that we must add as search roots must all be in blocks that dominate the early block (where the load can at earliest be scheduled).